### PR TITLE
Moving PRF to wolfcrypt

### PR DESCRIPTION
**Pull request by embhorn**
Migrate the pseudo random function component to the wolfcrypt library. Also created a compile time define that can be used to enable PRF, `WOLFSSL_HAVE_PRF`.

### CODE REVIEW

File: src/tls.c [hunk 1]
```diff
@@ -459,10 +174,15 @@ int BuildTlsFinished(WOLFSSL* ssl, Hashes* hashes, const byte* sender)
...
             side = tls_server;
 
-        ret = PRF((byte*)hashes, TLS_FINISHED_SZ, ssl->arrays->masterSecret,
+#ifndef NO_PRF
+        ret = wc_PRF((byte*)hashes, TLS_FINISHED_SZ, ssl->arrays->masterSecret,
                    SECRET_LEN, side, FINISHED_LABEL_SZ, handshake_hash, hashSz,
                    IsAtLeastTLSv1_2(ssl), ssl->specs.mac_algorithm,
                    ssl->heap, ssl->devId);
+#else
+        /* Pseudo random function must be enabled */
```
[the remark this answers is not in the document]
Comment (embhorn) [reply src/tls.c]: Got it.

**Comment by embhorn** [comment]
Retest this please.

**Comment by embhorn** [comment]
retest this please

**Comment by embhorn** [comment]
retest this please

**Comment by embhorn** [comment]
retest this please

File: src/tls.c [hunk 2]
```diff
@@ -555,8 +278,27 @@ static int _DeriveTlsKeys(byte* key_dig, word32 key_dig_len,
     XMEMCPY(seed,           sr, RAN_LEN);
     XMEMCPY(seed + RAN_LEN, cr, RAN_LEN);
 
-    ret = PRF(key_dig, key_dig_len, ms, msLen, key_label, KEY_LABEL_SZ,
+#ifdef WOLFSSL_HAVE_PRF
+    ret = wc_PRF_TLS(key_dig, key_dig_len, ms, msLen, key_label, KEY_LABEL_SZ,
                seed, SEED_LEN, tls1_2, hash_type, heap, devId);
+#else
+    /* Pseudo random function must be enabled */
+    ret = WOLFSSL_FATAL_ERROR;
```
Comment (toddouska) [inline src/tls.c]: We need a message here and shouldn't the return code be about not compile in?

File: src/tls.c [hunk 3]
```diff
@@ -628,8 +370,21 @@ static int _MakeTlsMasterSecret(byte* ms, word32 msLen,
     XMEMCPY(seed,           cr, RAN_LEN);
     XMEMCPY(seed + RAN_LEN, sr, RAN_LEN);
 
-    ret = PRF(ms, msLen, pms, pmsLen, master_label, MASTER_LABEL_SZ,
+#ifdef WOLFSSL_HAVE_PRF
+    ret = wc_PRF_TLS(ms, msLen, pms, pmsLen, master_label, MASTER_LABEL_SZ,
                seed, SEED_LEN, tls1_2, hash_type, heap, devId);
+#else
+    /* Pseudo random function must be enabled */
+    ret = WOLFSSL_FATAL_ERROR;
```
Comment (toddouska) [inline src/tls.c]: too generic, no message.

File: src/tls.c [hunk 4]
```diff
@@ -657,8 +412,26 @@ static int _MakeTlsExtendedMasterSecret(byte* ms, word32 msLen,
...
 {
-    return PRF(ms, msLen, pms, pmsLen, ext_master_label, EXT_MASTER_LABEL_SZ,
+    int ret;
+
+#ifdef WOLFSSL_HAVE_PRF
+    ret = wc_PRF_TLS(ms, msLen, pms, pmsLen, ext_master_label, EXT_MASTER_LABEL_SZ,
                sHash, sHashLen, tls1_2, hash_type, heap, devId);
+#else
+    /* Pseudo random function must be enabled */
+    ret = WOLFSSL_FATAL_ERROR;
```
Comment (toddouska) [inline src/tls.c]: too generic, no message.

File: src/tls.c [hunk 5]
```diff
@@ -786,10 +559,18 @@ int wolfSSL_make_eap_keys(WOLFSSL* ssl, void* msk, unsigned int len,
...
 
-    ret = PRF((byte*)msk, len, ssl->arrays->masterSecret, SECRET_LEN,
+#ifdef WOLFSSL_HAVE_PRF
+    ret = wc_PRF_TLS((byte*)msk, len, ssl->arrays->masterSecret, SECRET_LEN,
               (const byte *)label, (word32)XSTRLEN(label), seed, SEED_LEN,
               IsAtLeastTLSv1_2(ssl), ssl->specs.mac_algorithm,
               ssl->heap, ssl->devId);
+#else
+    /* Pseudo random function must be enabled */
+    ret = WOLFSSL_FATAL_ERROR;
```
Comment (toddouska) [inline src/tls.c]: too generic, no message.

File: wolfcrypt/src/hash.c [hunk 6]
```diff
@@ -928,3 +923,275 @@ int wc_HashFree(wc_HashAlg* hash, enum wc_HashType type)
...
+#ifdef WOLFSSL_HAVE_PRF
+
+#ifdef WOLFSSL_SHA384
+    #define P_HASH_MAX_SIZE WC_SHA384_DIGEST_SIZE
+#else
+    #define P_HASH_MAX_SIZE WC_SHA256_DIGEST_SIZE
+#endif
+
+/* compute p_hash for MD5, SHA-1, SHA-256, or SHA-384 */
+int wc_PRF(byte* result, word32 resLen, const byte* secret,
```
Comment (toddouska) [inline wolfcrypt/src/hash.c]: comment mentions p_hash but the name has changed.  This should be more consistent for clarity.

File: wolfcrypt/src/hash.c [hunk 7]
```diff
@@ -928,3 +923,275 @@ int wc_HashFree(wc_HashAlg* hash, enum wc_HashType type)
...
+
+#if defined(WOLFSSL_ASYNC_CRYPT) && !defined(WC_ASYNC_NO_HASH)
+    FREE_VAR(labelSeed, heap);
+#endif
+
+    return ret;
+}
+
+/* Wrapper for TLS 1.2 and TLSv1 cases to calculate PRF */
+/* In TLS 1.2 case call straight thru to p_hash */
```
Comment (toddouska) [inline wolfcrypt/src/hash.c]: p_hash no longer exists.

File: wolfcrypt/src/hash.c [hunk 8]
```diff
@@ -928,3 +923,275 @@ int wc_HashFree(wc_HashAlg* hash, enum wc_HashType type)
...
+            return BUFFER_E;
+
+        XMEMCPY(labelSeed, label, labLen);
+        XMEMCPY(labelSeed + labLen, seed, seedLen);
+
+        /* If a cipher suite wants an algorithm better than sha256, it
+         * should use better. */
+        if (hash_type < sha256_mac || hash_type == blake2b_mac)
+            hash_type = sha256_mac;
+        /* compute p_hash for MD5, SHA-1, SHA-256, or SHA-384 for TLSv1 PRF */
```
Comment (toddouska) [inline wolfcrypt/src/hash.c]: TLSv1 is wrong here.

Comment (toddouska) [reply wolfcrypt/src/hash.c]: Also, p_hash no longer exists.

**Comment by embhorn** [comment]
Hi @toddouska 
I updated the comments that referenced "p_hash" and added a new error for the missing PRF condition, and a message in `wolfSSL_ERR_reason_error_string()`.

File: wolfcrypt/src/hash.c [hunk 9]
```diff
@@ -1176,7 +1176,7 @@ int wc_PRF_TLS(byte* digest, word32 digLen, const byte* secret, word32 secLen,
          * should use better. */
         if (hash_type < sha256_mac || hash_type == blake2b_mac)
             hash_type = sha256_mac;
-        /* compute p_hash for MD5, SHA-1, SHA-256, or SHA-384 for TLSv1 PRF */
+        /* compute hash for MD5, SHA-1, SHA-256, or SHA-384 for TLSv1.2 PRF */
```
[the remark this answers is not in the document]
Comment (embhorn) [reply wolfcrypt/src/hash.c]: Got it.

File: src/tls.c [hunk 10]
```diff
@@ -180,8 +180,8 @@ int BuildTlsFinished(WOLFSSL* ssl, Hashes* hashes, const byte* sender)
                    IsAtLeastTLSv1_2(ssl), ssl->specs.mac_algorithm,
                    ssl->heap, ssl->devId);
 #else
-        /* Pseudo random function must be enabled */
-        ret = WOLFSSL_FATAL_ERROR;
+        /* Pseudo random function must be enabled in the configuration. */
+        ret = PRF_MISSING;
```
[the remark this answers is not in the document]
Comment (embhorn) [reply src/tls.c]: Updated all instances.

**Pull request merged**